File: src/main/java/com/temporal/agentic/engine/DagTopologyComputer.java
package com.temporal.agentic.engine;

import com.temporal.agentic.models.*;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Computes topological ordering and node readiness in a DAG.
 */
public class DagTopologyComputer {

    public Map<String, Integer> computeTopologicalLevels(DagSpec dagSpec) {
        Map<String, Integer> levels = new TreeMap<>();
        Map<String, Set<String>> incomingEdges = buildIncomingEdgesMap(dagSpec);
        Map<String, Set<String>> outgoingEdges = buildOutgoingEdgesMap(dagSpec);

        for (NodeDefinition node : dagSpec.getNodes()) {
            levels.put(node.getId(), 0);
        }

        boolean changed = true;
        while (changed) {
            changed = false;
            for (NodeDefinition node : dagSpec.getNodes()) {
                Set<String> incoming = incomingEdges.getOrDefault(node.getId(), new HashSet<>());
                int maxLevel = 0;
                for (String inNodeId : incoming) {
                    int inLevel = levels.getOrDefault(inNodeId, 0);
                    if (inLevel >= maxLevel) {
                        maxLevel = inLevel + 1;
                    }
                }
                int currentLevel = levels.get(node.getId());
                if (maxLevel > currentLevel) {
                    levels.put(node.getId(), maxLevel);
                    changed = true;
                }
            }
        }

        return levels;
    }

    public Set<String> getReadyNodes(DagSpec dagSpec, Map<String, String> nodeStatuses,
                                    WorkflowContext context, ConditionEvaluator evaluator) {
        Set<String> ready = new TreeSet<>();
        Map<String, Set<String>> incomingEdges = buildIncomingEdgesMap(dagSpec);

        for (NodeDefinition node : dagSpec.getNodes()) {
            if (nodeStatuses.getOrDefault(node.getId(), "PENDING").equals("PENDING")) {
                Set<String> incoming = incomingEdges.getOrDefault(node.getId(), new HashSet<>());

                boolean allPredecessorsComplete = true;
                for (String predId : incoming) {
                    String predStatus = nodeStatuses.getOrDefault(predId, "PENDING");
                    if (!predStatus.equals("SUCCESS") && !predStatus.equals("SKIPPED")) {
                        allPredecessorsComplete = false;
                        break;
                    }
                }

                if (allPredecessorsComplete) {
                    if (node.getCondition() == null || node.getCondition().trim().isEmpty() ||
                        evaluator.evaluate(node.getCondition(), context.getData())) {
                        ready.add(node.getId());
                    }
                }
            }
        }

        return ready;
    }

    private Map<String, Set<String>> buildIncomingEdgesMap(DagSpec dagSpec) {
        Map<String, Set<String>> map = new TreeMap<>();
        for (EdgeDefinition edge : dagSpec.getEdges()) {
            map.computeIfAbsent(edge.getTo(), k -> new TreeSet<>()).add(edge.getFrom());
        }
        return map;
    }

    private Map<String, Set<String>> buildOutgoingEdgesMap(DagSpec dagSpec) {
        Map<String, Set<String>> map = new TreeMap<>();
        for (EdgeDefinition edge : dagSpec.getEdges()) {
            map.computeIfAbsent(edge.getFrom(), k -> new TreeSet<>()).add(edge.getTo());
        }
        return map;
    }

    public List<String> computeDeterministicNodeOrder(Set<String> readyNodes) {
        return new ArrayList<>(readyNodes).stream()
                .sorted()
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/temporal/agentic/engine/DagTopologyComputer.kt
package com.temporal.agentic.engine

import com.temporal.agentic.models.DagSpec
import com.temporal.agentic.models.EdgeDefinition
import com.temporal.agentic.models.NodeDefinition
import com.temporal.agentic.models.WorkflowContext
import java.util.*

class DagTopologyComputer {
    
    fun computeTopologicalLevels(dagSpec: DagSpec): Map<String, Int> {
        val levels = TreeMap<String, Int>()
        val incomingEdges = buildIncomingEdgesMap(dagSpec)
        
        for (node in dagSpec.nodes) {
            levels[node.id!!] = 0
        }
        
        var changed = true
        while (changed) {
            changed = false
            for (node in dagSpec.nodes) {
                val incoming = incomingEdges[node.id] ?: emptySet()
                var maxLevel = 0
                for (inNodeId in incoming) {
                    val inLevel = levels[inNodeId] ?: 0
                    if (inLevel >= maxLevel) {
                        maxLevel = inLevel + 1
                    }
                }
                val currentLevel = levels[node.id]!!
                if (maxLevel > currentLevel) {
                    levels[node.id!!] = maxLevel
                    changed = true
                }
            }
        }
        
        return levels
    }
    
    fun getReadyNodes(
        dagSpec: DagSpec,
        nodeStatuses: Map<String, String>,
        context: WorkflowContext,
        evaluator: ConditionEvaluator
    ): Set<String> {
        val ready = TreeSet<String>()
        val incomingEdges = buildIncomingEdgesMap(dagSpec)
        
        for (node in dagSpec.nodes) {
            if (nodeStatuses[node.id] == "PENDING") {
                val incoming = incomingEdges[node.id] ?: emptySet()
                
                val allPredecessorsComplete = incoming.all { predId ->
                    val predStatus = nodeStatuses[predId] ?: "PENDING"
                    predStatus == "SUCCESS" || predStatus == "SKIPPED"
                }
                
                if (allPredecessorsComplete) {
                    val condition = node.condition
                    if (condition.isNullOrBlank() || evaluator.evaluate(condition, context.data)) {
                        ready.add(node.id!!)
                    }
                }
            }
        }
        
        return ready
    }
    
    private fun buildIncomingEdgesMap(dagSpec: DagSpec): Map<String, Set<String>> {
        val map = TreeMap<String, MutableSet<String>>()
        for (edge in dagSpec.edges) {
            map.computeIfAbsent(edge.to!!) { TreeSet() }.add(edge.from!!)
        }
        return map
    }
    
    private fun buildOutgoingEdgesMap(dagSpec: DagSpec): Map<String, Set<String>> {
        val map = TreeMap<String, MutableSet<String>>()
        for (edge in dagSpec.edges) {
            map.computeIfAbsent(edge.from!!) { TreeSet() }.add(edge.to!!)
        }
        return map
    }
    
    fun computeDeterministicNodeOrder(readyNodes: Set<String>): List<String> {
        return readyNodes.sorted()
    }
}
